main/main_timer_sync.cpp

```diff
@@ -172,9 +191,37 @@ MainFrameTime MainTimerSync::advance_checked(float p_frame_slice, int p_iteratio
 	// last clamping: make sure time_accum is between 0 and p_frame_slice for consistency between physics and idle
 	ret.clamp_idle(idle_minus_accum, idle_minus_accum + p_frame_slice);
 
+	// all the operations above may have turned ret.idle_step negative or zero, keep
+	// at least one tick
+	const float min_step = 1 / 1000000.0;
```

Comment:
Might be better to write as `1e-6`.

Reply:
@aaronfranke 1E-6 would be my preference as well, but the codebase has a different opinion.
```
$ grep "1E-6" * -R | wc -l
4
$ grep "1000000\.0" * -R | wc -l
30
```
The three uses of 1E-6 are all collision penetration tolerances, the fourth was from a binary, where all conversions from ticks to seconds I found are done in the ticks/1000000.0 form.

The SYNC_TIMER_DEBUG_ENABLED guards many runtime checks later in the code. The commented out line is there so if further problems appear, I can quickly enable the extra diagnostics again. They're not enabled in general even for debug builds for various reasons:
The warnings are just diagnostics that the new code has kicked in and prevented errors. The test case triggers them all, so in general, they're useless noise.
Some of the CRASH_CONDs have floating point comparisons in them, and you know how those are. They might trigger false positive crashes if they're just slightly violated (which still would be fine).
None of the CRASH_CONDs are of the 'we can't continue if this condition holds' type, they just indicate something has gone wrong in this part of the code that should be looked at.

I can rework those to only emit warnings in debug mode for everyone if the expected constraints are not met. That way, we get some diagnostics that something is going wrong that can be easily traced to the source without risking spurious crashes.

Reply:
Thinking again, taking one tick as the minimum output timestep here is wrong. A single tick is small enough to cause trouble in careless _process functions, for example if they use
`1 - cos(delta*omega)`
(which they should not!) with some omega in the range of one. A better minimal output timestep is probably a quarter of the input timestep.

Reply:
In practice I don't think you can guess in advance how users are going to use this delta (and thus there is no 'safe' value). I added something like this on a similar PR on a suggestion, but in retrospect really I think the best answer is to require all code to work when passed a zero delta.

A suggestion - we may be able to enforce this by passing a test zero delta at some point / points during runtime. This would ensure that any such bugs in core and client code were fixed sharpish.

Reply:
> A suggestion - we may be able to enforce this by passing a test zero delta at some point / points during runtime. This would ensure that any such bugs in core and client code were fixed sharpish.

That sounds like a good idea to me. Bad client code should fail fast. If code has problems with zero delta, it will likely also misbehave for very small delta. And worst case, the code runs into an error and aborts (only that one _process call). What is the expected result from that? No change, which is what is supposed to happen with zero delta.
I'd be for adding a _process(0) call once to any freshly added object, randomly as first to 100ths call or similar. Debug mode only, maybe. And then change the two bits of code that guarantee a minimal delta (yours in the basic timer query and mine here) to just guarantee non-negative delta.

Reply:
> A suggestion - we may be able to enforce this by passing a test zero delta at some point / points during runtime.

That's a very good idea.  These kinds of rare mistakes can be nightmarish to track down.